xclim/core/bootstrapping.py

```diff
@@ -0,0 +1,256 @@
+from dataclasses import dataclass
+from functools import wraps
+from inspect import signature
+from typing import Callable, Optional
+
+import numpy as np
+import xarray as xr
+from xarray.core.dataarray import DataArray
+
+import xclim.core.calendar as calendar
+from xclim.core.bootstrap_config import NO_BOOTSRAP, BootstrapConfig
+from xclim.core.calendar import percentile_doy
+
+
+def percentile_bootstrap(func):
+    """Decorator for indices which can be bootstrapped.
+
+    Only the percentile based indices may benefit from bootstrapping.
+
+    When the indices function is called with a BootstrapConfig parameter,
+    the decarator will take over the computation to iterate over the base period in this configuration.
+    @see compute_bootstrapped_exceedance_rate for the full Algorithm.
+
+    Example of declaration:
+    @declare_units(tas="[temperature]", t90="[temperature]")
+    @percentile_bootstrap
+    def tg90p(
+        tas: xarray.DataArray,
+        t90: xarray.DataArray,
+        freq: str = "YS",
+        bootstrap_config: BootstrapConfig = None,
+    ) -> xarray.DataArray:
+
+    Example when called:
+    >>> config = BootstrapConfig(percentile=90,
+                                percentile_window=5,
+                                in_base_slice=slice("2015-01-01", "2018-12-31"),
+                                out_of_base_slice=slice("2019-01-01", "2024-12-31"))
+    >>> tg90p(tas = da, t90=None, freq="MS", bootstrap_config=config)
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        bound_args = signature(func).bind(*args, **kwargs)
+        config = NO_BOOTSRAP
+        indice_window = None
+        for name, val in bound_args.arguments.items():
+            if name == "window":
+                indice_window = val
+            elif name == "freq":
+                freq = val
+            elif isinstance(val, DataArray):
+                da = val
```

Comment:
Won't this catch both tas and t90 ?

Reply:
Yes it could. In my tests I always explicitly put t90=None to avoid this.
I'm not quite sure how it could be handled better because the _name_ of the main parameter (here "tas") change between functions so I cannot rely on this.
We could check that there is **one and only one** DataArray in arguments when there is a bootstrap_config maybe ?

I feel like this makes the API a bit confusing to have some parameter excluding one another (t90 and bootstrap_config), but I don't see an other way of doing that with decorators.

Reply:
Actually it doesn't work anymore to do `t90=None`, since this commit beba80d5379d9d0aa4833341d3f8e93f878700ec I think. It gives this error now:
```
File "/.../indicator.py", line 621, in __call__
    self._update_attrs(
  File "/.../indicator.py", line 813, in _update_attrs
    attrs["cell_methods"] = merge_attributes(
  File "/.../formatting.py", line 264, in merge_attributes
    if attribute in in_ds.attrs or missing_str is not None:
AttributeError: 'NoneType' object has no attribute 'attrs'
```
Now, it works if I change the signature of tx90p to have a default None for t90:
```
@declare_units(tasmax="[temperature]", t90="[temperature]")
@percentile_bootstrap
def tx90p(
    tasmax: xarray.DataArray,
    t90: xarray.DataArray = None,
    freq: str = "YS",
    bootstrap_config: BootstrapConfig = NO_BOOTSRAP,
) -> xarray.DataArray:  # noqa: D401
```

Reply:
Hum. Ok, we'll try to think of something on our end and come back with a suggestion.